packages/prover/src/proof_provider/payload_store.ts

```diff
@@ -0,0 +1,154 @@
+    if (typeof blockId === "string" && blockId.startsWith("0x") && blockId.length === 64 + 2) {
+      return this.payloads.get(blockId);
+    }
+
+    // Given block id is a block number in hex
+    if (typeof blockId === "string" && blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(hexToNumber(blockId));
+    }
+
+    // Given block id is a block number in decimal string
+    if (typeof blockId === "string" && !blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(parseInt(blockId, 10));
+    }
+
+    // Given block id is a block number in decimal
+    if (typeof blockId === "number") {
+      return this.getOrFetchFinalizedPayload(blockId);
+    }
+
+    return undefined;
+  }
+
+  async getOrFetchFinalizedPayload(blockNumber: number): Promise<allForks.ExecutionPayload | undefined> {
+    if (blockNumber > this.finalizedRoots.max) {
+      throw new Error(
+        `Block number ${blockNumber} is higher than the latest finalized block number. We recommend to use block hash for unfinalized blocks.`
+      );
+    }
+
+    let blockELRoot = this.finalizedRoots.get(blockNumber);
+    if (!blockELRoot) {
+      const payloads = await getExecutionPayloadForBlockNumber(this.opts.api, this.finalizedRoots.min, blockNumber);
+      for (const payload of Object.values(payloads)) {
+        this.set(payload, true);
+      }
+    }
+
+    blockELRoot = this.finalizedRoots.get(blockNumber);
+    if (blockELRoot) {
+      return this.payloads.get(blockELRoot);
+    }
+
+    return undefined;
+  }
+
+  set(payload: allForks.ExecutionPayload, finalized: boolean): void {
+    const blockRoot = bufferToHex(payload.blockHash);
+    this.payloads.set(blockRoot, payload);
+    this.latestBlockRoot = blockRoot;
+
+    if (finalized) {
+      this.finalizedRoots.set(payload.blockNumber, blockRoot);
+    }
+  }
+
+  async processLCHeader(header: capella.LightClientHeader, finalized = false): Promise<void> {
+    const blockSlot = header.beacon.slot;
+    const blockNumber = header.execution.blockNumber;
+    const blockELRoot = bufferToHex(header.execution.blockHash);
+    const blockCLRoot = bufferToHex(header.beacon.stateRoot);
+    const existingELRoot = this.unfinalizedRoots.get(blockCLRoot);
+
+    if (finalized) {
+      if (existingELRoot) {
+        this.unfinalizedRoots.delete(blockCLRoot);
+      } else {
+        this.payloads.set(
+          bufferToHex(header.execution.blockHash),
+          (await getExecutionPayloads(this.opts.api, blockSlot, blockSlot))[blockSlot]
+        );
+      }
+      this.finalizedRoots.set(blockNumber, blockELRoot);
+
+      return;
+    }
+
+    if (existingELRoot && existingELRoot === blockELRoot) {
+      // We already have the payload for this block
+      return;
+    }
+
+    if (existingELRoot && existingELRoot !== blockELRoot) {
+      // Re-org happened, we need to update the payload
```

Comment:
If i understand correctly, I think el root by CL stateroot can never differ since stateRoot has executionPayloadHeader 's merkel hash included in it, if this happens, this is failure of consensus and is a critical error and you should just reject the header

Reply:
My understanding is, the reorg can happen on the CL and EL side. Each fork on CL can point to different fork on EL. So incase we receive a header update which for which we already have a payload in record but have different root that's a sign of such reorg. In that case we consider new version to be right.